### PR TITLE
WIP: Cleanup `LVM`

### PR DESCRIPTION
**Pull request by RussKie**
Fixes #

## Proposed changes

- WindowMessage had a duplicate list of `LVM_` messages. 
    * Rename the dups, update callsites to use the existing enum.
    * Add tests to confirm the correct behaviour
- Remove few redundant `else`s to reduce nesting

## Customer Impact

- Internal plumbing changes, should be not visible to consumers

## Regression? 

- No

## Risk

- Minimal, behaviour was verified with tests

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/2077)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs [hunk 1]
```diff
@@ -3279,72 +3279,69 @@ private unsafe ListViewItem FindItem(bool isTextSearch, string text, bool isPref
...
+                    // we can do this because SearchDirectionHint is set to the VK_*
+                    lvFindInfo.vkDirection = (uint)dir;
                 }
-            }
-            else
-            {
-                fixed (char *pText = text)
+                lvFindInfo.lParam = IntPtr.Zero;
+
+                int index = (int)SendMessage((int)LVM.FINDITEM,
```
Comment (hughbe) [inline src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: The casting is a bit gross - is there a way to try and avoid it? (generics etc?)

File: src/Common/src/Interop/Interop.ListViewMessages.cs [hunk 2]
```diff
@@ -11,6 +11,7 @@ internal static partial class Interop
     public enum LVM
```
Comment (hughbe) [inline src/Common/src/Interop/Interop.ListViewMessages.cs]: IMO this should be moved into ComCtl32 folder and namespaced. Also rename to Interop.LVM.cs

File: src/Common/src/Interop/Interop.ListViewMessages.cs [hunk 3]
```diff
@@ -11,6 +11,7 @@ internal static partial class Interop
...
     {
         FIRST                      = 0x1000,
+        GETBKCOLOR                 = FIRST + 0,
```
Comment (zsd4yr) [inline src/Common/src/Interop/Interop.ListViewMessages.cs]: We could honestly save the compiler some trouble by just doing the addition ourselves. Is is likely that FIRST would ever change?

Comment (weltkante) [reply src/Common/src/Interop/Interop.ListViewMessages.cs]: It can't change, programs compiled against older headers still need to be runnable.

Comment (zsd4yr) [reply src/Common/src/Interop/Interop.ListViewMessages.cs]: I just mean doing like
GETBKCOLOR = 0x1000,
SETBKCOLOR = 0x1001,
SETIMAGELIST = 0x1003,
...

Also, is GETBKCOLOR supposed to be a duplicate of FIRST and is there supposed to be no entry for 0x1002 ?

Comment (RussKie) [reply src/Common/src/Interop/Interop.ListViewMessages.cs]: This is how C++ defined and how a user likely to be searching for it. I don't like the idea of recalculating all values, or making it harder to people to search for consts.

File: src/Common/src/Interop/Interop.ListViewMessages.cs (hunk 2 above)
Comment (RussKie) [reply src/Common/src/Interop/Interop.ListViewMessages.cs]: Thanks, will do

File: src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs (hunk 1 above)
Comment (RussKie) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: We can probably tweak `SendMessage` method signature. I'll have a look

Comment (JeremyKuhne) [reply src/System.Windows.Forms/src/System/Windows/Forms/ListView.cs]: Why are you not calling `Interop` directly? That was a much cleaner call. We really should remove the internal helpers methods and add some sort of debug only method to `IHandle` for handling our `IsHandleCreated` check. We also need to create an issue to get rid of the rest of the `UnsafeNativeMethods.SendMessage` wraps.

File: src/Common/src/Interop/ComCtl32/Interop.HDITEM2.cs [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+internal static partial class Interop
+{
+    internal static partial class ComCtl32
+    {
+        // HDN_ITEMCHANGING will send us an HDITEM w/ pszText set to some random pointer.
+        // Marshal.PtrToStructure chokes when it has to convert a random pointer to a string.
+        // For HDN_ITEMCHANGING map pszText to an IntPtr
+        [StructLayout(LayoutKind.Sequential, CharSet = CharSet.Auto)]
+        public class HDITEM2
```
Comment (JeremyKuhne) [inline src/Common/src/Interop/ComCtl32/Interop.HDITEM2.cs]: I would change this to a struct, either `HDITEM` or `HDITEMW` and be explicit about the `CharSet` here, which is Unicode.

As it is defined currently, everything is blittable so it is an easy conversion. `NMHEADER` and `NMHDR` are also blittable as is and should be changed to structs. Then you can just use them as is, no `Marshal.PtrToStructure` needed.

Here is the definition:

``` C++
typedef struct _HD_ITEMW
{
    UINT    mask;
    int     cxy;
    LPWSTR  pszText;
    HBITMAP hbm;
    int     cchTextMax;
    int     fmt;
    LPARAM  lParam;
    int     iImage;        // index of bitmap in ImageList
    int     iOrder;
    UINT    type;           // [in] filter type (defined what pvFilter is a pointer to)
    void *  pvFilter;       // [in] fillter data see above
#if (NTDDI_VERSION >= NTDDI_VISTA)
    UINT   state;
#endif
} HDITEMW, *LPHDITEMW;
```

Which is:

``` C#
public struct HDITEMW
{
    public uint mask;
    public int cxy;
    public char* pszText; // Or IntPtr
    public IntPtr hbm;
    public int cchTextMax;
    public int fmt;
    public IntPtr lParam;
    public int iImage;
    public int iOrder;
    public uint type;
    public void* pvFilter; // Or IntPtr
    public uint state;
}
```

**Comment by hughbe** [comment]
Can be closed: https://github.com/dotnet/winforms/pull/2077

**Pull request closed without merging**